netcore/System.Private.CoreLib/src/System.Reflection/RuntimeAssembly.cs

```diff
@@ -137,7 +137,48 @@ internal static AssemblyName CreateAssemblyName (string assemblyString, out Runt
 		public override extern Type[] GetExportedTypes ();
 
 		[MethodImplAttribute (MethodImplOptions.InternalCall)]
-		public override extern Type[] GetForwardedTypes ();
+		internal extern Type[] GetTopLevelForwardedTypes ();
+
+		public override Type[] GetForwardedTypes () {
+			Type[] topLevelTypes = GetTopLevelForwardedTypes ();
+			List<Type> forwardedTypes = new List<Type> (topLevelTypes);
+			List<Exception> exceptions = new List<Exception> ();
+
+			foreach (Type t in topLevelTypes)
+				AddPublicNestedTypes (t, forwardedTypes, exceptions);
+
+			if (exceptions.Count > 0) {
+				forwardedTypes.AddRange (new Type[exceptions.Count]); // add one null Type for each exception
+				exceptions.InsertRange (0, new Exception[forwardedTypes.Count]); // align the Exceptions with the null Types
+				throw new ReflectionTypeLoadException (forwardedTypes.ToArray (), exceptions.ToArray ());
+			}
+
+			return forwardedTypes.ToArray ();
+		}
+
+		private static void AddPublicNestedTypes (Type type, List<Type> types, List<Exception> exceptions)
+		{
+			Type[] nestedTypes;
+
+			try {
+				nestedTypes = type.GetNestedTypes (BindingFlags.Public);
+			} 
+			catch (FileLoadException e) { exceptions.Add (e); return; }
+			catch (FileNotFoundException e) { exceptions.Add (e); return; }
+			catch (TypeLoadException e) { exceptions.Add (e); return; }
+			catch (IOException e) { exceptions.Add (e); return; }
+			catch (UnauthorizedAccessException e) { exceptions.Add (e); return; }
```

Comment:
maybe it's better to do:
```csharp
catch (Exception e) when (IsLoaderException(e)) { exceptions.Add (e); return; }
```

Reply:
I was told specifically not to do that because exception filters don't work properly on some platforms we support.

Reply:
That would use an exception filter which we are trying to avoid.

Reply:
ah, good to know